### DISCUSSION TITLE
Serverless Framework V3 End of Support - Now what?

### DISCUSSION THREAD

**Discussion by jphechter** [FearlessSolutions/engineering-practice-domains#20] [General]
Hey all,

As you may or may not be aware, Serverless Framework v3 will no longer be supported at the end of 2024. Users can fairly simply upgrade to v4, but for ["organizations earning over $2 million annually"](https://www.serverless.com/pricing) you will be required to purchase a paid subscription.

My current project is in support of an organization that is fairly heavily invested in Serverless on multiple applications. We're at a crossroads where we need to decide if we will stick with this service or migrate to something new. Are others facing this same issue, and if so what are you doing? Is what Serverless offers you worth paying?

-----
UPDATE 12/10:

The organization we're supporting has opted for the following. We're going to update to use Serverless v4 to ensure that no one will be without any potential security patches. For the most part, upgrading is not complicated and requires the least amount of development time. I'm expecting that everyone will have upgraded their respective repos by January. The team is still deciding if there is going to be a technology that everyone will use going forward, but we are also empowered to make our own decisions. Looking increasingly likely that we will use AWS CDK. We have no plans to move away from AWS in the near term, and our team uses JS/TS everywhere already. Using a tool that makes writing infrastructure code accessible to all of our devs is a huge plus.

**Comment by fearlessrob** [comment]
I love this question and I wish I knew more about `serverless` to help you answer it. Is this something that could potentially play out like Terraform / OpenTofu?

**Reply by jphechter** [reply]
Honestly, hard to say. My project is not _too_ heavily dependent on Serverless and we do already use OpenTofu. Migrated away from Terraform for largely similar reasons.

My team only works with one of the codebases that utilizes Serverless, and while there's probably not a one-size-fits-all solution here we have been included in the conversation as trusted thought partners. Some of the folks have already discussed upgrading and accepting the costs for a short time because there are some known issues in v3 that have been patched in v4. For my project though it's likely that we'll just continue with v3 even past the end of support for just a brief time while we migrate away.

**Comment by asharonbaltazar** [comment]
Our codebase [moved over](https://github.com/Enterprise-CMCS/macpro-mako/pull/698/commits/e8990ac6906b8fc28d03dea51991035900abd77c) from Serverless to AWS CDK. I'm not close to being an expert but I'd say we've had a positive and (to my recollection) issue-less transition. Some of the positives: 

- cut our AWS bill "literally" in half. It's crazy
- made AWS a lot more accessible to the team since it uses JS/TS
- adding to the previous point, removed the incomprehensible YAML configuration files that made no sense to the layman

TLDR: very positive

**Reply by jphechter** [reply]
This is probably the most likely thing we will do. There's probably going to be a mix of increased dependence on OpenTofu and using AWS CDK. We're almost exclusively building in an AWS environment, so while it's "nice" to be cloud agnostic it's unlikely that we'll be changing any time soon. As I understand it, Serverless v4 is no longer cloud agnostic either. So no matter how you spin it, you're locking into a single cloud provider.

We're also a TypeScript team. Making this part of our infrastructure more approachable to the rest of the team would be huge.

Hadn't considered any cost savings, but I'm curious if that's something we'd benefit from. The majority of our billing costs are actually tied up in Route53 (long story there), but any dollar saved is appreciated.

**Comment by cgodwin1** [comment]
We are also in the process of transitioning [eRegs](https://github.com/Enterprise-CMCS/cmcs-eregulations) from Serverless to CDK. We're attempting to perform an incremental transition since our app is broken out into separate CloudFormation stacks; the site+db, static assets, the eCFR & Federal Register parsers, the text extractor, and a few other minor components.

So far, we like what we see. We are already seeing a benefit in that CDK does not require a deployment bucket for each stack, which was an issue with Serverless that prevented us from being able to have the number of ephemeral deploys that we want at any given time.

Also, we frequently had issues with Serverless not properly removing ephemeral stacks. Usually a bucket did not get emptied, or a log group failed to delete, etc., requiring manual intervention/cleanup with AWS CLI or the console. We tried to fix this a few times but it never fully went away: see [here](https://github.com/Enterprise-CMCS/cmcs-eregulations/pull/794), [here](https://github.com/Enterprise-CMCS/cmcs-eregulations/pull/1033), and [here](https://github.com/Enterprise-CMCS/cmcs-eregulations/pull/1260) for some of our attempted fixes. We have not seen issues like that with CDK yet. Even better that CDK supports automatically deleting objects from S3 when destroying the stack.

I'll go against the grain and say I actually prefer Serverless's YAML configuration files. They mostly make sense to me and require a bit less boilerplate in our codebase at least. But, I can already see the flexibility that CDK and TypeScript offer, so it's a worthwhile tradeoff in my opinion.

Other aspects we've had to consider and/or rework so far:
- We had to add a lot of boilerplate to ensure IAM roles, boundary permissions, etc. are set up properly.
- Deploy stages; native to Serverless but have to be manually implemented in CDK.
- Trying to get CDK to take ownership of our existing Serverless-deployed stacks. We ideally want to transition without tearing down prod and deploying from scratch, but we have a plan for a switchover as well. We haven't gotten there yet so we'll see how it plays out.
- We made use of Serverless plugins quite a bit, for example `serverless-wsgi` for Django, `serverless-cloudfront-invalidate`, and `serverless-s3-sync`. We'll need to handle that ourselves now or find alternatives.
    - Upside to this: Some of the plugins were not well-maintained. We [had to take over](https://github.com/Enterprise-CMCS/cmcs-eregulations/pull/1326) `cloudfront-invalidate` because of a vulnerability in vm2 that was never fixed by the developer. `serverless-python-requirements` is also [causing headaches for us](https://github.com/serverless/serverless-python-requirements/issues/846).
- [We tried](https://github.com/Enterprise-CMCS/cmcs-eregulations/compare/5a2c2c30e36944751959f701eb6298abb11c789a...2514-localstack-migration) to use Serverless with LocalStack earlier this year and it was a painful experience, half of the features didn't work and several plugins were outright broken locally. If we ever try LocalStack again, it would be interesting to see if CDK makes life easier.

Overall a positive experience, but definitely requires more manual work and sometimes some new ways of thinking about infrastructure.

**Reply by jphechter** [reply]
Glad to hear about the overall positive experience. Gives me some hope that this won't be a overwhelmingly painful process. I'm not really sure what our biggest stumbling points will be. One of the plugins we rely on pretty heavily is `serverless-dynamodb`. Haven't gotten around to considering what a replacement will look like yet.

Best of luck in the continued migration process. I'll be very curious to hear more as you guys march forward. Excited to compare notes when we start making moves over here.
